staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller_debug.go

```diff
@@ -0,0 +1,199 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/server/mux"
+)
+
+func (cfgCtl *configController) Install(c *mux.PathRecorderMux) {
+	// TODO(yue9944882): handle "Accept" header properly
+	// debugging dumps a CSV content for three levels of granularity
+	// 1. row per priority-level
+	c.UnlistedHandleFunc("/debug/flowcontrol/dump_per_prioritylevel", cfgCtl.dumpPerPriorityLevel)
+	// 2. row per queue
+	c.UnlistedHandleFunc("/debug/flowcontrol/dump_per_queue", cfgCtl.dumpPerQueue)
+	// 3. row per request
+	c.UnlistedHandleFunc("/debug/flowcontrol/dump_per_request", cfgCtl.dumpPerRequest)
+}
+
+func (cfgCtl *configController) dumpPerPriorityLevel(w http.ResponseWriter, r *http.Request) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	tabWriter := tabwriter.NewWriter(w, 8, 0, 1, ' ', 0)
+	columnHeaders := []string{
+		"PriorityLevelName",
+		"FlowSchemaName",
+		"ActiveQueues",
+		"IsIdle",
+		"RobinIndex",
+		"WaitingRequests",
+		"ExecutingRequests",
+	}
+	tabPrint(tabWriter, rowForHeaders(columnHeaders))
+	for _, plState := range cfgCtl.priorityLevelStates {
+		if plState.queues == nil {
+			continue
+		}
+		queueSetDigest := plState.queues.Dump()
+		activeQueueNum := 0
+		for _, q := range queueSetDigest.Queues {
+			if len(q.Requests) > 0 {
+				activeQueueNum++
+			}
+		}
+
+		tabPrint(tabWriter, rowForPriorityLevel(
+			plState.pl.Name,
+			activeQueueNum,
+			plState.queues.IsIdle(),
+			queueSetDigest.RobinIndex,
+			queueSetDigest.Waiting,
+			queueSetDigest.Executing,
+		))
+	}
+	runtime.HandleError(tabWriter.Flush())
+}
+
+func (cfgCtl *configController) dumpPerQueue(w http.ResponseWriter, r *http.Request) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	tabWriter := tabwriter.NewWriter(w, 8, 0, 1, ' ', 0)
+	columnHeaders := []string{
+		"PriorityLevelName",
+		"PendingRequests",
+		"ExecutingRequests",
+		"RobinIndex",
+		"Index",
+		"Length",
+		"VirtualStart",
+	}
+	tabPrint(tabWriter, rowForHeaders(columnHeaders))
+	for _, plState := range cfgCtl.priorityLevelStates {
+		if plState.queues == nil {
+			continue
+		}
+		queueSetDigest := plState.queues.Dump()
+		if plState.queues == nil {
+			continue
+		}
+		for _, q := range queueSetDigest.Queues {
+			tabPrint(tabWriter, rowForQueue(
+				plState.pl.Name,
+				queueSetDigest.Waiting,
+				queueSetDigest.Executing,
+				queueSetDigest.RobinIndex,
```

Comment:
1. I do not favor denormalizing the data.  Each column of the per-queue rows should be specific to the row's queue.

2. Each queue has its own number of executing requests; this row should include that number.

Reply:
> 1. I do not favor denormalizing the data. Each column of the per-queue rows should be specific to the row's queue.

removed the denormalized columns. PTAL

> 2. Each queue has its own number of executing requests; this row should include that number.

agree. added the columns